src/background.c

```diff
@@ -0,0 +1,817 @@
+    return NO;
+  else if (strcmp (permissions[0], "ask") == 0)
+    return ASK;
+  else
+    {
+      g_autofree char *a = g_strjoinv (" ", (char **)permissions);
+      g_warning ("Wrong permission format, ignoring (%s)", a);
+    }
+
+  return UNSET;
+}
+
+static Permission
+get_permission (const char *app_id)
+{
+  g_autoptr(GVariant) perms = NULL;
+
+  perms = get_all_permissions ();
+  if (perms)
+    return get_one_permission (app_id, perms);
+
+  return UNSET;
+}
+
+static void
+set_permission (const char *app_id,
+                Permission permission)
+{
+  g_autoptr(GError) error = NULL;
+  const char *permissions[2];
+
+  if (permission == ASK)
+    permissions[0] = "ask";
+  else if (permission == YES)
+    permissions[0] = "yes";
+  else if (permission == NO)
+    permissions[0] = "no";
+  else
+    {
+      g_warning ("Wrong permission format, ignoring");
+      return;
+    }
+  permissions[1] = NULL;
+
+  if (!xdp_impl_permission_store_call_set_permission_sync (get_permission_store (),
+                                                           PERMISSION_TABLE,
+                                                           TRUE,
+                                                           PERMISSION_ID,
+                                                           app_id,
+                                                           (const char * const*)permissions,
+                                                           NULL,
+                                                           &error))
+    {
+      g_dbus_error_strip_remote_error (error);
+      g_warning ("Error updating permission store: %s", error->message);
+    }
+}
+
+typedef enum {
+  AUTOSTART_FLAGS_NONE        = 0,
+  AUTOSTART_FLAGS_ACTIVATABLE = 1 << 0,
+} AutostartFlags;
+
+static void
+handle_request_background_in_thread_func (GTask *task,
+                                          gpointer source_object,
+                                          gpointer task_data,
+                                          GCancellable *cancellable)
+{
+  Request *request = (Request *)task_data;
+  GVariant *options;
+  const char *app_id;
+  Permission permission;
+  const char *reason = NULL;
+  gboolean autostart_requested = FALSE;
+  gboolean autostart_enabled;
+  gboolean allowed;
+  g_autoptr(GError) error = NULL;
+  const char * const *autostart_exec = { NULL };
+  AutostartFlags autostart_flags = AUTOSTART_FLAGS_NONE;
+  gboolean activatable = FALSE;
+  g_auto(GStrv) commandline = NULL;
+
+  REQUEST_AUTOLOCK (request);
+
+  options = (GVariant *)g_object_get_data (G_OBJECT (request), "options");
+  g_variant_lookup (options, "reason", "&s", &reason);
+  g_variant_lookup (options, "autostart", "b", &autostart_requested);
+  g_variant_lookup (options, "commandline", "^a&s", &autostart_exec);
+  g_variant_lookup (options, "dbus-activatable", "b", &activatable);
+
+  if (activatable)
+    autostart_flags |= AUTOSTART_FLAGS_ACTIVATABLE;
```

Comment:
Should we fail on unknown flags?

Reply:
we validate options when we receive them from the caller. These here are already validated.